include/swift/Runtime/Mutex.h

```diff
@@ -186,121 +114,687 @@ class Mutex {
+  ///   object synchronize-with this operation. unlock() does not synchronize
+  ///   with a failed try_writeLock().
+  /// - The behavior is undefined if the calling thread already owns
+  ///   the read or write lock (likely a deadlock)?
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  bool try_writeLock() {
+    return ReadWriteLockPlatformHelper::try_writeLock(Handle);
+  }
+
+  /// The writeUnlock() method has the following properties:
+  /// - Behaves as an atomic operation.
+  /// - Releases the calling thread's ownership of the write lock
+  ///   and synchronizes-with the subsequent successful lock operations on
+  ///   the same object.
+  /// - The behavior is undefined if the calling thread does not own
+  ///   the write lock.
+  /// - Does not throw exceptions but will halt on error (fatalError).
+  void writeUnlock() { ReadWriteLockPlatformHelper::writeUnlock(Handle); }
+
+  /// Acquires read lock before calling the supplied critical section and
+  /// releases lock on return from critical section. Callers must not mutate
+  /// the data protected by the ReadWriteLock while holding the read lock, the
+  /// write lock must be used.
+  ///
+  /// This call can block while waiting for the lock to become available.
+  ///
+  /// For example the following reads the cached value while holding
+  /// the read lock.
+  ///
+  /// ```
+  ///   rw.readLock([&value] { value = cachedValue; });
+  /// ```
+  ///
+  /// Precondition: ReadWriteLock not held by this thread, undefined otherwise.
+  template <typename CriticalSection>
+  void readLock(CriticalSection criticalSection) {
+    readLock();
+    criticalSection();
+    readUnlock();
+  }
+
+  /// Acquires write lock before calling the supplied critical section and
+  /// releases lock on return from critical section.
+  ///
+  /// This call can block while waiting for the lock to become available.
+  ///
+  /// For example the following updates the cached value while holding
+  /// the write lock.
+  ///
+  /// ```
+  ///   rw.writeLock([&newValue] { cachedValue = newValue });
+  /// ```
+  ///
+  /// Precondition: ReadWriteLock not held by this thread, undefined otherwise.
+  template <typename CriticalSection>
+  void writeLock(CriticalSection criticalSection) {
+    writeLock();
     criticalSection();
-    condition.notifyAll();
+    writeUnlock();
+  }
+
+  /// Acquires read lock before calling the read critical section and if
+  /// the critical section returns `true` (done) readWriteLock returns to the
+  /// caller. If the read critical section returns `false` the read lock
+  /// will be dropped, the write lock will be acquired and then the read
+  /// critical section will be called again. Then if the critical section
+  /// returns `true` (done) readWriteLock returns to the caller. If the read
+  /// critical section returns `false` then the write critical section will
+  /// be called (with write lock still held).
+  ///
```

Comment:
This function is too fancy and only used once as Joe noted below. Please delete it.